### ISSUE TITLE
Support SameSite cookie (Strict, Lax, None)

### ISSUE THREAD

**Issue by kiennguyen** [playframework/play1#1328]
### Are you looking for help?

From Google Chrome 80, they are planning to released from february, 2020, they will force SameSite=None for all Cookies if we want to use for third-party context. It seems to be reversed the result as before (SameSite=Lax by default). Please see: https://www.chromium.org/updates/same-site
Therefore, we need to update our cookies with declarative setting SameSite from Playframework Controller.

### Play Version (1.5.x / etc)
Playframework 1.5.0

### Expected Behavior
1. Support SameSite enum inside the Http.Cookie
2. SameSite enum supports three values: Strict, Lax and None

**Comment by jvosloo** [comment]
How exactly would you do that in a controller?

**Comment by miller11** [comment]
@jvosloo I think what @kiennguyen is getting at is just extending the cookie interface to allow us to set the SameSite attribute just like we can currently set other cookie attributes like expiration or httpOnly.

**Comment by jvosloo** [comment]
Ah ok, I see in Play2 the default seems to be "lax": https://www.playframework.com/documentation/2.8.x/SettingsSession

Would that be a sensible default for Play1 as well then?

**Comment by miller11** [comment]
That is an interesting choice. I would think that not setting it all would be best default since that is the current behavior and it mimics things like expiration and httpOnly. Another thing to consider is not all browsers support the SameSite attribute so setting it `lax` by default may cause issues with some browsers.

**Comment by jvosloo** [comment]
I hear you. Check out this discussion: https://github.com/playframework/playframework/issues/7316

**Comment by kiennguyen** [comment]
Thanks @miller11 . Yes, we just want to have possibility to set SameSite property like httpOnly or maxAge in Http.Cookie class. No need to have default value.

**Comment by kiennguyen** [comment]
@jvosloo Is that clear? Do you know the plan to have this feature in the play 1.5.x?

**Comment by jvosloo** [comment]
@kiennguyen makes sense to me yes.  I'm not not one of the committers, so can't answer on the plan. Perhaps @miller11 has an idea

**Comment by miller11** [comment]
Also not a committer, however I don't think this is a simple move forward. Play uses the [Cookie interface](https://netty.io/4.1/api/index.html) from Netty to set cookies as seen in the [PlayHandler](https://github.com/playframework/play1/blob/master/framework/src/play/server/PlayHandler.java#L365). As you can see even in the newest 4.1 release of Netty they don't yet support the SameSite attribute (Play 1.5.3 is currently on [Netty 3.10.6.Final](https://github.com/playframework/play1/blob/master/framework/dependencies.yml#L70) . 

There is an [issue](https://github.com/netty/netty/issues/8530) in the Netty repo and even a [PR](https://github.com/eclipse-vertx/vert.x/pull/3202) for SameSite support, but seems like there is some discussion when this would get merged. 

One annoying issue to note about the SameSite attribute is that it doesn't seem to be respected uniformly [across browsers](https://itnext.io/user-agent-sniffing-only-way-to-deal-with-upcoming-samesite-cookie-changes-6f79a18e541) yet.

**Comment by tomparle** [comment]
Thanks everyone for your involvement and suggestions so far !
It seems that Netty will only include this in version 5.x and up for compatibility reasons.
Regarding migration from Netty 3, I did some progress could not complete it : https://github.com/playframework/play1/issues/1284 <= some help will be needed here because Play core's code was starting to be too tricky for me, unfortunately.

**Comment by xabolcs** [comment]
Netty 4 support is almost ready to ship.

netty/netty#10050

**Comment by dumedeiros** [comment]
Alguém já sabe informar como adicionar o SameSite?

**Comment by dumedeiros** [comment]
Does anyone already know how to add SameSite?

**Comment by Alexandermjos** [comment]
Hi all.
Is anyone working on samesite for play 1.x? 

I have a working version / custom build based on Play 1.5.2 which has support for samesite

@miller11 Play classes is extending netty classes, so netty doesnt need to support samesite?

**Comment by tazmaniax** [comment]
@Alexandermjos very interested to see what you have, can you create a pull request?

**Comment by Alexandermjos** [comment]
@tazmaniax I have created a pull request. I ended up starting from scratch and not using the implementation I had based on Play 1.5.2, more info regarding that in the PR. Looking forward for feedback 😃  I have not added any as reviewers. Should I add someone, or do people of interests add themselves? This is my first PR to an open source project, so not sure if I did everything correct. 🤞

**Comment by tazmaniax** [comment]
@Alexandermjos looks good. I can see there was some discussion above regarding what the default value should be and the consensus seemed to be that there shouldn't be any (aka "normal"). My preference would be "lax" like Play2 so without thinking there is some automatic protection - this is what "lax" was designed to be, a reasonable compromise that in most cases won't break anything. I realise this means people will need to consider carefully before upgrading to a version with this fix but I think in respect to security that is only a good thing.

**Comment by asolntsev** [comment]
@tazmaniax if I am not mistaken, modern browsers behave like "lax" by default. If it's true, it's good enough to leave "SameSite" attribute empty by default.

**Comment by tazmaniax** [comment]
@asolntsev right, I think I came to the same conclusion 😄

**Comment by tazmaniax** [comment]
@asolntsev from @asolntsev's point and what I found, I don't think it's necessary to explicitly set the default to "lax" because browsers now just assume "lax" when SameSite is not defined. It would only be useful if there was a browser that didn't support the latest version of the spec but I think this is a very rare case so I think it's good to leave it to what you have now. Thanks!

**Comment by Alexandermjos** [comment]
I have committed some changes which @asolntsev requested 😄 
Also added the space before '(' @tazmaniax 👍

**Comment by Alexandermjos** [comment]
@xael-fry Any thoughts? 😃

**Comment by lebort** [comment]
Any progress on this issue, @xael-fry or @asolntsev? Could the pull request soon be merged, such that the issue can be fixed in an upcoming release?
